--- before/configure.ac  2022-05-19 17:06:36.921274500 +0200
+++ after/configure.ac   2022-05-17 11:51:17.319667600 +0200
@@ -308,14 +308,14 @@


 # Decide if we want to build the tests or not.
-buildTestUtilities=false
-if test "x$YAJL_FOUND" = "x1"; then
+# buildTestUtilities=false
+# if test "x$YAJL_FOUND" = "x1"; then
     # Regression tests will not be able to run without the logging support.
     # But we still have the unit tests.
     # if test "$debugLogs" = "true"; then
-        buildTestUtilities=true
+#        buildTestUtilities=true
     # fi
-fi
+# fi


 AM_CONDITIONAL([TEST_UTILITIES], [test $buildTestUtilities = true])
