--- mmdb.lua	2022-04-04 09:32:41.456286600 +0200
+++ mmdb2.lua	2022-04-04 09:33:25.016286600 +0200
@@ -166,7 +166,7 @@
                             MMDB_entry_data_s *const entry_data,
                             ...);
 ]])
-local lib = ffi.load("libmaxminddb")
+local lib = ffi.load("/usr/share/bunkerweb/deps/lib/libmaxminddb.so")
 local consume_map, consume_array
 local consume_value
 consume_value = function(current)
